File: llm_generated_code/ChatGPT/CPP/Level2/Q11.cpp
#include <iostream>
#include <vector>
using namespace std;

void dfs(vector<vector<int>>& map, int x, int y) {
    int rows = map.size();
    int cols = map[0].size();
    if (x < 0 || y < 0 || x >= rows || y >= cols || map[x][y] != 1) {
        return;
    }

    map[x][y] = -1;  // Mark this cell as visited by setting it to -1

    // Explore all four directions
    dfs(map, x + 1, y);
    dfs(map, x - 1, y);
    dfs(map, x, y + 1);
    dfs(map, x, y - 1);
}

int countEnclosedLandCells(vector<vector<int>>& map) {
    int rows = map.size();
    int cols = map[0].size();

    // Mark all land cells connected to the boundary as -1
    for (int i = 0; i < rows; i++) {
        if (map[i][0] == 1) dfs(map, i, 0);
        if (map[i][cols - 1] == 1) dfs(map, i, cols - 1);
    }
    for (int j = 0; j < cols; j++) {
        if (map[0][j] == 1) dfs(map, 0, j);
        if (map[rows - 1][j] == 1) dfs(map, rows - 1, j);
    }

    // Count the remaining land cells that are enclosed
    int enclosedCount = 0;
    for (int i = 1; i < rows - 1; i++) {
        for (int j = 1; j < cols - 1; j++) {
            if (map[i][j] == 1) {
                enclosedCount++;
            }
        }
    }

    return enclosedCount;
}

int main() {
    vector<vector<int>> map = {
        {0, 0, 0, 0, 0, 0},
        {0, 1, 1, 0, 1, 0},
        {0, 1, 0, 0, 1, 0},
        {0, 0, 0, 1, 0, 0},
        {0, 1, 1, 1, 0, 0},
        {0, 0, 0, 0, 0, 0}
    };

    int result = countEnclosedLandCells(map);

    cout << "Number of enclosed land cells: " << result << endl;

    return 0;
}

File: llm_generated_code/ChatGPT/Python/Level2/Q11.py
def dfs_mark(grid, x, y, rows, cols):
    if x < 0 or x >= rows or y < 0 or y >= cols or grid[x][y] != 1:
        return
    # Mark this cell as visited by changing it to -1
    grid[x][y] = -1
    # Perform DFS in all 4 directions
    dfs_mark(grid, x - 1, y, rows, cols)  # Up
    dfs_mark(grid, x + 1, y, rows, cols)  # Down
    dfs_mark(grid, x, y - 1, rows, cols)  # Left
    dfs_mark(grid, x, y + 1, rows, cols)  # Right

def count_surrounded_land_cells(grid):
    rows, cols = len(grid), len(grid[0])

    # Mark all land cells connected to the boundary as non-surrounded
    for r in range(rows):
        for c in [0, cols - 1]:  # Only left and right boundaries
            if grid[r][c] == 1:
                dfs_mark(grid, r, c, rows, cols)
    for c in range(cols):
        for r in [0, rows - 1]:  # Only top and bottom boundaries
            if grid[r][c] == 1:
                dfs_mark(grid, r, c, rows, cols)

    # Count cells that are still 1 (completely surrounded land cells)
    surrounded_count = sum(1 for r in range(1, rows - 1) for c in range(1, cols - 1) if grid[r][c] == 1)

    return surrounded_count

# Example usage
archipelago_map = [
    [0, 0, 0, 0, 0],
    [0, 1, 1, 0, 0],
    [0, 1, 0, 1, 0],
    [0, 0, 1, 0, 0],
    [0, 0, 0, 0, 0]
]

surrounded_land_cells = count_surrounded_land_cells(archipelago_map)
print("Number of completely surrounded land cells:", surrounded_land_cells)

